### ISSUE TITLE
crio: Unable to pull images form internal registry

### ISSUE THREAD

**Issue by kameshsampath** [kubernetes/minikube#10171]
```
minikube version: v1.16.0
commit: 9f1e482427589ff8451c4723b6ba53bb9742fbb1
```

**Steps to reproduce the issue:** 

1. Start Minikube as 
```shell
minikube start -p $PROFILE_NAME \
  --memory=$MEMORY --cpus=$CPUS \
  --disk-size=50g \
  --insecure-registry='10.0.0.0/24'
```
2.  Apply the registry and registry-aliases add-on 
3. Build and push an image to the Minikube registry as `example.com/demo/greeter`
4.  Deploy a pod using the command
```shell
kubectl run demo-greeter -n tektontutorial \
 --generator='run-pod/v1' \
 --image='example.com/demo/greeter' && \
kubectl expose pod demo-greeter -n tektontutorial --port 8080 --type=NodePort
```

The kubectl get events shows the crashed pod with the following logs:

**Full output of failed command:** 
```
0s          Warning   Failed      pod/demo-greeter                        Failed to pull image "registry.minikube/rhdevelopers/tekton-tutorial-greeter": rpc error: code = Unknown desc = error pinging docker registry registry.minikube: Get https://registry.minikube/v2/: dial tcp 10.101.166.84:443: connect: connection refused
```

With `--insecure-registry` specified, the pod still tries to resolve the url as `https` than `http`

CC: @afbjorklund

**Comment by kameshsampath** [comment]
some more contextual info 

`minikube ssh -- cat /etc/hosts`

```
127.0.0.1       localhost
127.0.1.1 tektontutorial
192.168.64.1    host.minikube.internal
192.168.64.20   control-plane.minikube.internal
10.101.166.84   example.org
10.101.166.84   example.com
10.101.166.84   test.com
10.101.166.84   test.org
10.101.166.84   registry.minikube
```

`kubectl get svc -n kube-system`

```
AME       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   10.96.0.10      <none>        53/UDP,53/TCP,9153/TCP   79m
registry   ClusterIP   *10.101.166.84*   <none>        80/TCP,443/TCP           78m
```

`kubectl get cm coredns -n kube-system -o yaml | yq eval '.data' - `

```
Corefile: |-
  .:53 {
      errors
       rewrite name example.org  registry.kube-system.svc.cluster.local
  rewrite name example.com  registry.kube-system.svc.cluster.local
  rewrite name test.com  registry.kube-system.svc.cluster.local
  rewrite name test.org  registry.kube-system.svc.cluster.local
  rewrite name registry.minikube  registry.kube-system.svc.cluster.local

      health {
         lameduck 5s
      }
      ready
      kubernetes cluster.local in-addr.arpa ip6.arpa {
         pods insecure
         fallthrough in-addr.arpa ip6.arpa
         ttl 30
      }
      prometheus :9153
      forward . /etc/resolv.conf {
         max_concurrent 1000
      }
      cache 30
      loop
      reload
      loadbalance
  }

```

**Comment by afbjorklund** [comment]
Worked here.

```
  Type     Reason     Age              From               Message
  ----     ------     ----             ----               -------
  Normal   Scheduled  4s               default-scheduler  Successfully assigned default/busybox to minikube
  Normal   Pulled     4s               kubelet            Successfully pulled image "registry.minikube/busybox" in 16.245797ms
  Normal   Pulling    3s (x2 over 4s)  kubelet            Pulling image "registry.minikube/busybox"
  Normal   Created    3s (x2 over 4s)  kubelet            Created container busybox
  Normal   Started    3s (x2 over 3s)  kubelet            Started container busybox
  Normal   Pulled     3s               kubelet            Successfully pulled image "registry.minikube/busybox" in 18.747139ms
  Warning  BackOff    1s (x2 over 2s)  kubelet            Back-off restarting failed container
```

You might want to double-check `minikube ssh docker info`:

```
 Insecure Registries:
  10.96.0.0/12
  10.0.0.0/24
  127.0.0.0/8
```

**Comment by kameshsampath** [comment]
that sound weird ! was it on macOS ?

Let me try again on a fresh instance, do you see any issue with minikube start command ??

**Comment by kameshsampath** [comment]
after a bit of debug @afbjorklund , this happens only with `container-runtime=cri-o`, if I use `container-runtime=docker` then all works well. Wondering something to do with internal work on `cri-o`?

**Comment by medyagh** [comment]
@kameshsampath interesting do we still have this issue on latest minikube ? I wonder if this issue happens if we specify a different CNI ?

**Comment by kameshsampath** [comment]
still get the same issue with crio, not sure what CNI driver I can use on macOS.
________________________________

**Comment by ricardozanini** [comment]
Just for your information, I got this issue today with the latest Minikube version. Same setup MacOS, podman, and cri-o. I also started the cluster with the flag `--insecure-registry=10.0.0.0/24`. The kubelet controller insists on using HTTPS instead of HTTP.

The problem is that the registry service is created exposing the ports `80:5000` and  `443:443`. But the registry pod, only exposes 5000. So obviously, kubelet or any client will be able to pull from 443.

I think this is a bug @kameshsampath @medyagh 

Cheers!

**Comment by medyagh** [comment]
> he latest Minikube version. Same setup MacOS, podman, and cri-o. I also started the cluster with the flag `--insecure-registry=10.0.0.0/24`. The kubelet controller insists on using HTTPS instead of HTTP.
> 
> The problem is that the registry service is created exposing the ports `80:5000` and `443:443`. But the registry pod, only exposes 5000. So obviously, kubelet or any client won't be able to pull from 443.

[ricardozanini](https://github.com/ricardozanini)
would you be interested to make a PR to fix this ?

of the issue is Container needs to open a non 443 port you can do that in the Kic Package

**Comment by ricardozanini** [comment]
@medyagh you bet I would! :)

I'll take a look later today and send a PR.
